test/csuite/wt3363_checkpoint_op_races/main.c

```diff
@@ -0,0 +1,498 @@
+/*-
+ * Public Domain 2014-2017 MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * JIRA ticket reference: WT-3362
+ *
+ * Test case description: There are a number of operations that we run that we
+ * expect not to conflict with or block against a running checkpoint. This test
+ * aims to run repeated checkpoints in a thread, while running an assortment
+ * of operations that we expect to execute quickly on further threads. To
+ * ensure that we catch any blockages we introduce a very large delay into the
+ * checkpoint and measure that no operation takes 1/2 the length of this delay.
+ *
+ * Failure mode: We monitor the execution time of all operations and if we find
+ * any operation taking longer than 1/2 the delay time, we abort dumping a core
+ * file which can be used to determine what operation was blocked.
+ */
+
+int handle_error(WT_EVENT_HANDLER *, WT_SESSION *, int, const char *);
+int handle_message(WT_EVENT_HANDLER *, WT_SESSION *, const char *);
+void* monitor(void *);
+void* do_checkpoints(void *);
+void* do_ops(void *);
+void op_bulk(WT_CONNECTION *);
+void op_bulk_unique(WT_CONNECTION *, int, int);
+void op_create(WT_CONNECTION *);
+void op_create_unique(WT_CONNECTION *, int, int);
+void op_cursor(WT_CONNECTION *);
+void op_drop(WT_CONNECTION *, int);
+
+#define	N_THREADS 10
+#define	RUNTIME 900.0
```

Comment:
It can take some time for the underlying issues to surface (specifically those in WT-3362) as it was an alignment of operations executing at the right time. The worst case I saw was just over 5 minutes before we saw a reproduction. I made this 15 minutes to honor the "three times" rule of reproducing a bug.